File: Leetcode/1416 Restore The Array.ts
const MOD = 1e9 + 7;

function numberOfArrays(s: string, k: number): number {
    const cache = new Array<number>(s.length).fill(-1);
    function dfs(i: number): number {
        if (i === s.length) {
            return 1;
        }
        if (cache[i] !== -1) {
            return cache[i];
        }
        if (s[i] === '0') {
            return cache[i] = 0;
        }
        let element = 0, nWays = 0;
        for (let j = i; j < s.length; j++) {
            element = 10 * element + parseInt(s[j]);
            if (element > k) {
                break;
            }
            nWays = (nWays + dfs(j + 1)) % MOD;
        }
        return cache[i] = nWays;
    }
    return dfs(0);
}

File: Leetcode/1416 Restore The Array.py
MOD = 1_000_000_007

class Solution:
    def numberOfArrays(self, s, k):
        cache = [None] * len(s)
        def dfs(i):
            if i == len(s):
                return 1
            if cache[i] is not None:
                return cache[i]
            elem, cache[i] = 0, 0
            for j in range(i, len(s)):
                elem = elem * 10 + int(s[j])
                if not (1 <= elem <= k):
                    break
                cache[i] = (cache[i] + dfs(j + 1)) % MOD
            return cache[i]
        return dfs(0)
